### ISSUE TITLE
exportAsImage does not work for pie and donut chart if its model is setShowDataLabels(true) - Primefaces 5.3

### ISSUE THREAD

**Issue by java4me** [primefaces/primefaces#920]
This issue happens in primefaces 5.3:
........
&lt;h:form id="formPreview" prependId="false" enctype="multipart/form-data">
    &lt;p:commandButton type="button" value="Export" icon="ui-icon-extlink" onclick="exportDemoChart()"/>
    &lt;br/>
        &lt;p:chart type="pie" model="#{chartView.pieModel1}" style="width:400px;height:300px" widgetVar="demoBar"/>
        &lt;p:dialog id="dlgDemoChart" widgetVar="dlgDemoChart" showEffect="slide" resizable="false" appendToBody="true" header="Export">
    &lt;p:outputPanel id="demoChartImgOutput" layout="block" style="width:500px;height:300px"/>
&lt;/p:dialog>  

&lt;script type="text/javascript">  
    function exportDemoChart() {  
        //export image  
        $('#demoChartImgOutput').empty().append(PF('demoBar').exportAsImage());  
        //show the dialog  
        PF('dlgDemoChart').show();  
    }  
&lt;/script>  
&lt;/h:form>

..........

```
   public PieChartModel getPieModel1() {
    if (pieModel1 == null) {
        pieModel1 = new PieChartModel();

        pieModel1.set("Brand 1", 540);
        pieModel1.set("Brand 2", 325);
        pieModel1.set("Brand 3", 702);
        pieModel1.set("Brand 4", 421);

        pieModel1.setTitle("Simple Pie");
        pieModel1.setLegendPosition("w");
        pieModel1.setShowDataLabels(true);
    }
    return pieModel1;
}
```

**Comment by kukel** [comment]
Does it work using plain jqplot?

**Comment by cagataycivici** [comment]
If it also doesn't work with jqplot, then I' have to close as won't fix.

**Comment by kukel** [comment]
no response in a month, please close

**Comment by jrobertsz66** [comment]
did you guys get this to work?  I also cannot get this to work wit a donut chart.

**Comment by cagataycivici** [comment]
Does it work using plain jqplot?

**Comment by larskn** [comment]
I found out that this is not an issue in PrimeFaces, but in jqPlot.

When rendering the image, jqPlot searches for the css classes `jqplot-table-legend-swatch-outline` and `jqplot-table-legend-label`. But those are simply not rendered by the `$.jqplot.DonutLegendRenderer`.

I created an issue in the jqPlot project: https://github.com/jqPlot/jqPlot/issues/78

I also created a workaround that seems to fix the issue by setting the required classes to the legend.

``` javascript
function fixLegendForImageRendering() {
    $("#my-donut-chart tr.jqplot-table-legend").each(function (index, legendRow) {
        $(legendRow).children("td.jqplot-table-legend").each(function (index, legendElement) {
            if (index % 2 == 0) {
                var swatchElement = $(legendElement)
                    .addClass("jqplot-table-legend-swatch");
                var swatchOutline = swatchElement.children("div")
                    .addClass("jqplot-table-legend-swatch-outline");
                var swatch = swatchOutline.children("div.jqplot-table-legend-swatch");
                swatch.css("background-color", swatch.css("borderTopColor"));
            } else {
                $(legendElement).addClass("jqplot-table-legend-label");
            }
        });
    });
}
```

It can be used via the jqPlot postDraw hook:

``` javascript
$.jqplot.postDrawHooks.push(fixLegendForImageRendering);
```

**Comment by tandraschko** [comment]
We can upgrade jqplot in the future if the issue has been fixed.

**Comment by edudoda** [comment]
This bug will be fixed in next version of jQplot (1.0.10):
Donut chart legend is not exported correctly because of missing css classes id=78
